react-components/src/data/search/queryBuilder/Clause.js

```diff
@@ -0,0 +1,156 @@
+import React, { Fragment } from "react";
+
+import Clauses from "./Clauses";
+import ids from "../../search/ids";
+
+import { build, FormSelectField } from "@cyverse-de/ui-lib";
+import { Field, FieldArray, getIn } from "formik";
+import MenuItem from "@material-ui/core/MenuItem";
+
+function ClauseSelector(props) {
+    const {
+        field: { name },
+        form: { values },
+        root,
+        collaboratorsUtil,
+        presenter,
+        parentId,
+    } = props;
+
+    let typeFieldName = name ? `${name}.type` : "type";
+    let argsFieldName = name ? `${name}.args` : "args";
+
+    let selectedClauseName = getIn(values, typeFieldName);
+
+    return (
+        <Fragment>
+            <ClauseType
+                typeFieldName={typeFieldName}
+                argsFieldName={argsFieldName}
+                root={root}
+                parentId={parentId}
+            />
+            <ClauseArgs
+                selectedClauseName={selectedClauseName}
+                argsFieldName={argsFieldName}
+                root={root}
+                collaboratorsUtil={collaboratorsUtil}
+                presenter={presenter}
+                parentId={parentId}
+            />
+        </Fragment>
+    );
+}
+
+function ClauseType(props) {
+    const { typeFieldName, argsFieldName, root, parentId } = props;
+
+    return (
+        <Field
+            name={typeFieldName}
+            render={({ field: { onChange, ...field }, ...props }) => {
+                return (
+                    <FormSelectField
+                        {...props}
+                        id={build(parentId, ids.conditionSelector)}
+                        fullWidth={false}
+                        field={field}
+                        onChange={(event) => {
+                            resetOnClauseChange(
+                                event,
+                                typeFieldName,
+                                argsFieldName,
+                                props.form
+                            );
+                            onChange(event);
+                        }}
+                    >
+                        {ClauseOptions(root, parentId)}
+                    </FormSelectField>
+                );
+            }}
+        />
+    );
+}
+
+function ClauseArgs(props) {
+    const {
+        selectedClauseName,
+        argsFieldName,
+        root,
+        collaboratorsUtil,
+        presenter,
+        parentId,
+    } = props;
+
+    let clauseObj = Clauses.componentMap[selectedClauseName];
+    let isGroupType = isGroup(selectedClauseName);
+
+    let ClauseComponent = clauseObj && clauseObj.component;
+
+    if (ClauseComponent) {
+        let FieldType = isGroupType ? FieldArray : Field;
```

Comment:
Thanks!!  After so many months removed from writing it, I realized how hard it was to follow so I tried to put more effort into that.  Formik does make it much easier since they allow you at any point to check the current state and they give you helper functions to change values when needed.
Your comment makes me realize I should probably go back and add more comments though to add some clarity, so I'll do that before merging.